File: plugins/plugin-browser/typescript/src/actions/click.ts
import type {
  Action,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { ServiceType } from "@elizaos/core";
import type { BrowserService } from "../services/browser-service.js";
import {
  ActionError,
  handleBrowserError,
  ServiceNotAvailableError,
  SessionError,
} from "../utils/errors.js";

export const browserClickAction: Action = {
  name: "BROWSER_CLICK",
  similes: ["CLICK_ELEMENT", "TAP", "PRESS_BUTTON"],
  description: "Click on an element on the webpage",

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const browserEnabled =
      runtime.getSetting("ENABLE_BROWSER") === "true" ||
      runtime.getSetting("BROWSER_ENABLED") === "true";

    if (!browserEnabled) {
      return false;
    }

    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      return false;
    }

    const text = message.content?.text?.toLowerCase() ?? "";
    return text.includes("click") || text.includes("tap") || text.includes("press");
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult | undefined> => {
    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      const error = new ServiceNotAvailableError();
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "click on element"
      );
      return {
        text: "Browser service is not available",
        success: false,
        data: {
          actionName: "BROWSER_CLICK",
          error: "service_not_available",
        },
        values: {
          success: false,
          errorType: "service_not_available",
        },
      };
    }

    const session = await service.getOrCreateSession();
    if (!session) {
      const error = new SessionError("No active browser session");
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "click on element"
      );
      return {
        text: "No active browser session",
        success: false,
        data: {
          actionName: "BROWSER_CLICK",
          error: "no_session",
        },
        values: {
          success: false,
          errorType: "no_session",
        },
      };
    }

    const text = message.content?.text ?? "";
    const match = text.match(/click (?:on |the )?(.+)$/i);
    const description = match ? match[1] : "element";

    const result = await service.getClient().click(session.id, description);
    if (!result.success) {
      throw new ActionError("click", description, new Error(result.error ?? "Click failed"));
    }

    const responseContent: Content = {
      text: `I've successfully clicked on "${description}"`,
      actions: ["BROWSER_CLICK"],
      source: message.content?.source ?? "action",
    };

    await callback?.(responseContent);

    return {
      text: responseContent.text ?? "",
      success: true,
      data: {
        actionName: "BROWSER_CLICK",
        element: description,
        sessionId: session.id,
      },
      values: {
        success: true,
        element: description,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user}}",
        content: { text: "Click on the search button" },
      },
      {
        name: "{{agent}}",
        content: {
          text: "I've clicked on the search button.",
          actions: ["BROWSER_CLICK"],
        },
      },
    ],
  ],
};

File: plugins/plugin-browser/rust/src/actions/click.rs
use crate::services::BrowserService;
use crate::types::ActionResult;
use crate::utils::{action_error, parse_click_target, session_error};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::error;

pub const CLICK_ACTION_NAME: &str = "BROWSER_CLICK";
pub const CLICK_SIMILES: &[&str] = &["CLICK_ELEMENT", "TAP", "PRESS_BUTTON"];
pub const CLICK_DESCRIPTION: &str = "Click on an element on the webpage";

pub async fn browser_click(service: Arc<BrowserService>, message: &str) -> ActionResult {
    let session = match service.get_or_create_session().await {
        Ok(s) => s,
        Err(e) => {
            let err = session_error(e);
            return ActionResult::failure(err.user_message);
        }
    };

    let description = parse_click_target(message);

    let client = service.get_client();
    let result = client.click(&session.id, &description).await;

    match result {
        Ok(response) if response.success => {
            let mut data = HashMap::new();
            data.insert(
                "actionName".to_string(),
                serde_json::json!(CLICK_ACTION_NAME),
            );
            data.insert("element".to_string(), serde_json::json!(description));
            data.insert("sessionId".to_string(), serde_json::json!(session.id));

            ActionResult::success(data)
        }
        Ok(response) => {
            let err = action_error("click", &description, response.error.as_deref());
            ActionResult::failure(err.user_message)
        }
        Err(e) => {
            error!("Error in BROWSER_CLICK action: {}", e);
            let err = action_error("click", &description, Some(&e));
            ActionResult::failure(err.user_message)
        }
    }
}
